Sort AAE differences before acting upon them via read-repair

slfritchie opened the pull request:
```
        %% Sort the keys.  For vnodes that use backends that preserve
        %% lexicographic sort order for BKeys, this is a big
        %% improvement.  For backends that do not, e.g. Bitcask, sorting
        %% by BKey is unlikely to be any worse.  For Riak CS's use
        %% pattern, sorting may have some benefit since block N is
        %% likely to be adjacent on disk to block N+1.
```

slfritchie committed:
Sort AAE differences before acting upon them via read-repair

engelsanchez commented on src/hashtree.erl:
```diff
@@ -117,6 +117,7 @@
          local_compare/2,
          compare/2,
          compare/3,
+         compare/4,
```
I would be in favor of reducing the number of convenience variations of this function, but that can be done later. Notice that in the documentation at the beginning of the file compare/2 and compare/3 are explicitly named.

jrwest replied:
ooh, i just realized. This file is not in `riak_kv` any more. It has been in core since https://github.com/basho/riak_kv/pull/652

jrwest replied:
another note, more pertinent to the original comment. Adding `compare/4`  would be useful for simplifying some of the code in `hashtree_tree.erl`.

engelsanchez replied:
BTW I like compare/4. I would keep it and eye the others for cleanup. Again, not pressing now.

engelsanchez commented on src/riak_kv_index_hashtree.erl:
```diff
@@ -600,12 +608,8 @@ do_compare(Id, Remote, AccFun, From, State) ->
             spawn_link(fun() ->
                                Remote(init, self()),
-                               Result = case AccFun of
-                                            undefined ->
```
Notice how compare/2 passes undefined to compare/3 as the AccFun. That would make it use hashtree:compare/2, which defaults to an accumulator function that concatenates lists. Also notice how compare/2 is not used anywhere in the Riak tree. If something were to happen to compare/2, I would not tell anybody.  If someone was even more caremad and trimmed all these different "convenient" versions of compare through these modules, I wouldn't say anything either.

engelsanchez commented on src/riak_kv_exchange_fsm.erl:
```diff
@@ -186,27 +206,53 @@ key_exchange(timeout, State=#state{local=LocalVN,
+                     lists:foldl(fun({DiffReason, BKeyBin}, Count) ->
+                                         {B, K} = binary_to_term(BKeyBin),
+                                         T = {B, K, DiffReason},
+                                         if Count rem 5000 == 0 ->
```
It would be nice to have this batching parameter as a macro a the top of the file

engelsanchez commented on src/riak_kv_exchange_fsm.erl:
```diff
@@ -171,11 +171,31 @@ key_exchange(timeout, State=#state{local=LocalVN,
+    LogFile2 = lists:flatten(io_lib:format("~s/~s/out.~p.~p.~p",
+                                           [TmpDir, ?MODULE, NA, NB, NC])),
+    ok = filelib:ensure_dir(LogFile2),
+    Remote = fun(get_bucket, {L, B}=_X) ->
```
_X here doesn't really do anything.

engelsanchez commented:
I'm done going over the code now. The comments I had were very minor, with the only little issue being the inconsistency of riak_kv_index_hashtree:compare/2 and the handling of the undefined AccFun (just getting rid of compare/2 would do).  I'll be testing this tomorrow.

engelsanchez commented:
And of course, the other "minor" issue of a file having completely moved out of the repo that Jordan mentioned :)

slfritchie committed:
Address review comments

The only Dialyzer differences between the base of the branch
and the tip is the line numbers for the complaints that
are present.

slfritchie committed:
Fix a few dialyzer complaints

slfritchie committed:
Remove hashtree:compare/2 and hashtree:compare/3 functions.

These functions are only used by tests and by developer
scaffolding.  See also:
https://github.com/basho/riak_kv/pull/646#discussion_r6673687

slfritchie commented:
I think I've addressed all of the review comments.  Dialyzer appears happy enough: the total number of warnings has dropped, and no new warnings have been added.

slfritchie commented:
Oh, oops.  So I patched a hashtree.erl that's now a ghost?

slfritchie committed:
Derp, there's a hashtree_eqc.erl

engelsanchez commented:
@slfritchie yes, we should do that simple hashtree patch in a separate PR in its new home in riak_core and update this. After that and a bit of testing on the final code, assuming it's all good, I should be able to give you a quick +1

slfritchie commented:
Engel: https://github.com/basho/riak_core/pull/411

engelsanchez commented:
I believe the comments have been addressed. I've verified sorted AAE repairs manually. EUnit tests have only failures that I've observed lately in the develop branch (get_put_monitor_eqc, crdts) but that don't seem to happen in the builder machines. It looks good. :+1: :dancer: 

I'm going to push a manual merge that I used for testing that takes care of the moved hashtree files.
